discovery/ec2/ec2.go

```diff
@@ -146,17 +145,33 @@ func NewDiscovery(conf *SDConfig, logger log.Logger) *Discovery {
 	if logger == nil {
 		logger = log.NewNopLogger()
 	}
-	d := &Discovery{
-		aws: &aws.Config{
+
+	sess, err := session.NewSessionWithOptions(session.Options{
+		Config: aws.Config{
 			Endpoint:    &conf.Endpoint,
 			Region:      &conf.Region,
 			Credentials: creds,
 		},
-		profile:  conf.Profile,
-		roleARN:  conf.RoleARN,
+		Profile: conf.Profile,
+	})
+	if err != nil {
+		return nil
```

Comment:
Hm @brian-brazil now I wonder if there would be a better way to pass an error from here to the outside. Maybe wrap the above call to `session.NewSessionWithOptions` in [`session.Must(...)`](https://docs.aws.amazon.com/sdk-for-go/api/aws/session/#Must) so that the process exits with an error if it fails?

Reply:
No, we shouldn't be causing all of your monitoring to fall over because one SD is having issues. What we should do here is try again on each refresh.

Reply:
Would you like to send a PR to fix that?

Reply:
Yeah, I'll create a follow-up PR next week.

Reply:
I just checked the implementation of `session.NewSessionWithOptions` again and it appears to me that this can only fail if there are some local configuration issues (like invalid AWS config file or specifying a profile which doesn't exist in the config etc.), but this shouldn't fail due to any external dependencies. Would it be ok to error out in the initialization in this case (like I suggested initially, I guess returning an error from `NewDiscoverer` could work) or do you still want this to be retried?

Reply:
Hmm, we should probably still retry in case such a problem is fixed at runtime.

Reply:
(Draft) PR is here: https://github.com/prometheus/prometheus/pull/8356 It compiles, but didn't have the chance to test this yet. I'll see if I can get this done next week.